File: app/src/main/java/com/iita/akilimo/views/activities/DatesActivity.java
package com.iita.akilimo.views.activities;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.RadioGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.SwitchCompat;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.FragmentManager;

import com.crashlytics.android.Crashlytics;
import com.google.android.gms.common.util.Strings;
import com.iita.akilimo.R;
import com.iita.akilimo.dao.AppDatabase;
import com.iita.akilimo.databinding.ActivityDatesBinding;
import com.iita.akilimo.entities.AdviceStatus;
import com.iita.akilimo.entities.ScheduledDate;
import com.iita.akilimo.inherit.BaseActivity;
import com.iita.akilimo.utils.DateHelper;
import com.iita.akilimo.utils.Tools;
import com.iita.akilimo.utils.enums.EnumAdviceTasks;
import com.iita.akilimo.views.fragments.dialog.DateDialogPickerFragment;

import org.joda.time.LocalDate;

public class DatesActivity extends BaseActivity {
    Toolbar toolbar;

    RelativeLayout lytPlantingHarvest;

    AppCompatButton btnPickPlantingDate;
    AppCompatButton btnPickHarvestDate;
    AppCompatButton btnCancel;
    AppCompatButton btnFinish;

    TextView lblSelectedPlantingDate;

    TextView lblSelectedHarvestDate;


    RadioGroup rdgAlternativeDate;

    RadioGroup rdgPlantingWindow;

    RadioGroup rdgHarvestWindow;

    SwitchCompat flexiblePlanting;
    SwitchCompat flexibleHarvest;

    ActivityDatesBinding binding;

    String selectedPlantingDate;
    String selectedHarvestDate;
    int plantingWindow = 0;
    int harvestWindow = 0;
    boolean alternativeDate;
    boolean alreadyPlanted;

    ScheduledDate scheduledDate;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityDatesBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());


        context = this;
        database = AppDatabase.getDatabase(context);

        //set widgets
        toolbar = binding.toolbar;
        lytPlantingHarvest = binding.lytPlantingHarvest;
        btnPickPlantingDate = binding.btnPickPlantingDate;
        btnPickHarvestDate = binding.btnPickHarvestDate;
        btnCancel = binding.twoButtons.btnCancel;
        btnFinish = binding.twoButtons.btnFinish;
        lblSelectedPlantingDate = binding.lblSelectedPlantingDate;
        lblSelectedHarvestDate = binding.lblSelectedHarvestDate;
        rdgAlternativeDate = binding.rdgAlternativeDate;
        rdgPlantingWindow = binding.rdgPlantingWindow;
        rdgHarvestWindow = binding.rdgHarvestWindow;
        flexiblePlanting = binding.flexiblePlanting;
        flexibleHarvest = binding.flexibleHarvest;


        initToolbar();
        initComponent();
    }

    @Override
    protected void initToolbar() {
        toolbar.setNavigationIcon(R.drawable.ic_left_arrow);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(getString(R.string.lbl_planting_harvest_dates));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        toolbar.setNavigationOnClickListener(v -> validate(true));
    }

    @Override
    protected void initComponent() {
        //now for title pickers
        btnPickPlantingDate.setOnClickListener(view -> dialogDatePickerLight(true, false));
        btnPickHarvestDate.setOnClickListener(view -> dialogDatePickerLight(false, true));
        btnFinish.setOnClickListener(view -> validate(false));
        btnCancel.setOnClickListener(view -> closeActivity(false));

        flexiblePlanting.setOnCheckedChangeListener((compoundButton, isChecked) -> {
            rdgPlantingWindow.clearCheck();
            plantingWindow = 0;
            if (isChecked) {
                if (alreadyPlanted) {
                    showCustomWarningDialog(getString(R.string.lbl_already_planted_title), getString(R.string.lbl_already_planted_text));
                    flexiblePlanting.setChecked(false);
                } else {
                    rdgPlantingWindow.setVisibility(View.VISIBLE);
                }
            } else {
                rdgPlantingWindow.setVisibility(View.GONE);
            }
        });

        flexibleHarvest.setOnCheckedChangeListener((compoundButton, isChecked) -> {
            rdgHarvestWindow.setVisibility(View.GONE);
            rdgHarvestWindow.clearCheck();
            harvestWindow = 0;
            if (isChecked) {
                rdgHarvestWindow.setVisibility(View.VISIBLE);
            } else {
                rdgHarvestWindow.setVisibility(View.GONE);
            }
        });

        rdgPlantingWindow.setOnCheckedChangeListener((radioGroup, radioIndex) -> {
            switch (radioIndex) {
                case R.id.rdPlantingOneMonth:
                    plantingWindow = 1;
                    break;
                case R.id.rdPlantingTwoMonths:
                    plantingWindow = 2;
                    break;
                default:
                    plantingWindow = 0;
                    break;
            }
        });
        rdgHarvestWindow.setOnCheckedChangeListener((radioGroup, radioIndex) -> {
            switch (radioIndex) {
                case R.id.rdHarvestOneMonth:
                    harvestWindow = 1;
                    break;
                case R.id.rdHarvestTwoMonths:
                    harvestWindow = 2;
                    break;
                default:
                    harvestWindow = 0;
                    break;
            }
        });

        rdgAlternativeDate.setOnCheckedChangeListener((radioGroup, radioIndex) -> {
            switch (radioIndex) {
                case R.id.rdYes:
                    lytPlantingHarvest.setVisibility(View.VISIBLE);
                    alternativeDate = true;
                    break;
                case R.id.rdNo:
                    lytPlantingHarvest.setVisibility(View.GONE);
                    alternativeDate = false;
                    break;
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        scheduledDate = database.scheduleDateDao().findOne();

        if (scheduledDate != null) {
            alternativeDate = scheduledDate.getAlternativeDate();
            String pd = scheduledDate.getPlantingDate();
            String hd = scheduledDate.getHarvestDate();
            int pw = scheduledDate.getPlantingWindow();
            int hw = scheduledDate.getHarvestWindow();

            alreadyPlanted = scheduledDate.getAlreadyPlanted();

            DateHelper.dateTimeFormat = "dd/MM/yyyy";
            LocalDate pDate = DateHelper.formatToLocalDate(pd);
            LocalDate hDate = DateHelper.formatToLocalDate(hd);
            lblSelectedPlantingDate.setText(pDate.toString());
            lblSelectedHarvestDate.setText(hDate.toString());
            rdgAlternativeDate.check(alternativeDate ? R.id.rdYes : R.id.rdNo);

            if (pw > 0) {
                flexiblePlanting.setChecked(true);
                rdgPlantingWindow.check(pw == 1 ? R.id.rdPlantingOneMonth : R.id.rdPlantingTwoMonths);
            }
            if (hw > 0) {
                flexibleHarvest.setChecked(true);
                rdgHarvestWindow.check(hw == 1 ? R.id.rdHarvestOneMonth : R.id.rdHarvestTwoMonths);
            }
            //assign these values to the global parameters
            selectedHarvestDate = hd;
            selectedPlantingDate = pd;
        }
    }

    @Override
    protected void validate(boolean backPressed) {
        if (Strings.isEmptyOrWhitespace(selectedPlantingDate)) {
            showCustomWarningDialog(getString(R.string.lbl_invalid_planting_date), getString(R.string.lbl_planting_date_prompt));
            return;
        }

        if (Strings.isEmptyOrWhitespace(selectedHarvestDate)) {
            showCustomWarningDialog(getString(R.string.lbl_invalid_harvest_date), getString(R.string.lbl_harvest_date_prompt));
            return;
        }

        try {
            if (scheduledDate == null) {
                scheduledDate = new ScheduledDate();
            }

            alreadyPlanted = DateHelper.olderThanCurrent(selectedPlantingDate);
            scheduledDate.setHarvestDate(selectedHarvestDate);
            scheduledDate.setHarvestWindow(harvestWindow);
            scheduledDate.setPlantingDate(selectedPlantingDate);
            scheduledDate.setPlantingWindow(plantingWindow);
            scheduledDate.setAlternativeDate(alternativeDate);
            scheduledDate.setAlreadyPlanted(alreadyPlanted);

            database.scheduleDateDao().insert(scheduledDate);

            database.adviceStatusDao().insert(new AdviceStatus(EnumAdviceTasks.PLANTING_AND_HARVEST.name(), true));

            closeActivity(backPressed);

        } catch (Exception ex) {
            Toast.makeText(context, ex.getMessage(), Toast.LENGTH_SHORT).show();
            Crashlytics.log(Log.ERROR, LOG_TAG, ex.getMessage());
            Crashlytics.logException(ex);
        }
    }

    private void dialogDatePickerLight(boolean pickPlantingDate, boolean pickHarvestDate) {
        final FragmentManager fm = getSupportFragmentManager();

        DateDialogPickerFragment pickerFragment = new DateDialogPickerFragment(true);
        if (pickHarvestDate) {
            pickerFragment = new DateDialogPickerFragment(true, selectedPlantingDate);
        }
        pickerFragment.show(fm, DateDialogPickerFragment.TAG);

        pickerFragment.setOnDismissListener((myCalendar, selectedDate, plantingDateSelected, harvestDateSelected) -> {
            long date_ship_millis = myCalendar.getTimeInMillis();
            if (pickPlantingDate) {
                selectedPlantingDate = DateHelper.getSimpleDateFormatter().format(myCalendar.getTime());
                lblSelectedPlantingDate.setText(Tools.formatLongToDateString(date_ship_millis));
                selectedHarvestDate = null;
                lblSelectedHarvestDate.setText(null);
                alreadyPlanted = DateHelper.olderThanCurrent(selectedPlantingDate);
                if (alreadyPlanted) {
                    flexiblePlanting.setChecked(false);
                }
            } else if (pickHarvestDate) {
                selectedHarvestDate = DateHelper.getSimpleDateFormatter().format(myCalendar.getTime());
                lblSelectedHarvestDate.setText(Tools.formatLongToDateString(date_ship_millis));
            }
        });

    }
}

File: app/src/main/java/com/akilimo/mobile/views/activities/DatesActivity.kt
package com.akilimo.mobile.views.activities

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.activity.viewModels
import com.akilimo.mobile.R
import com.akilimo.mobile.databinding.ActivityDatesBinding
import com.akilimo.mobile.inherit.BindBaseActivity
import com.akilimo.mobile.interfaces.IDatePickerDismissListener
import com.akilimo.mobile.utils.DateHelper.formatToLocalDate
import com.akilimo.mobile.viewmodels.DatesViewModel
import com.akilimo.mobile.viewmodels.factory.DatesViewModelFactory
import com.akilimo.mobile.views.fragments.dialog.DateDialogPickerFragment
import io.sentry.Sentry
import java.util.Calendar

class DatesActivity : BindBaseActivity<ActivityDatesBinding>() {

    private val viewModel: DatesViewModel by viewModels {
        DatesViewModelFactory(application = this.application)
    }

    override fun inflateBinding() = ActivityDatesBinding.inflate(layoutInflater)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setupToolbar(binding.toolbar, R.string.lbl_planting_harvest_dates) { validate(true) }
        observeViewModel()
        setupUI()
    }

    override fun onResume() {
        super.onResume()
        viewModel.loadSchedule()
    }

    private fun observeViewModel() {
        viewModel.plantingDate.observe(this) { updatePlantingDate(it) }
        viewModel.harvestDate.observe(this) { updateHarvestDate(it) }
        viewModel.alternativeDate.observe(this) { updateAlternativeDate(it) }
        viewModel.plantingWindow.observe(this) { updatePlantingWindow(it) }
        viewModel.harvestWindow.observe(this) { updateHarvestWindow(it) }
    }

    private fun setupUI() {
        setupPlantingUI()
        setupHarvestUI()
        setupAlternativeDateUI()
        setupButtons()
    }

    private fun updatePlantingDate(dateStr: String?) {
        binding.cardPlanting.lblSelectedPlantingDate.text =
            formatToLocalDate(dateStr ?: "", "dd-MM-yyyy").toString()
    }

    private fun updateHarvestDate(dateStr: String?) {
        binding.cardHarvest.lblSelectedHarvestDate.text =
            formatToLocalDate(dateStr ?: "", "dd-MM-yyyy").toString()
    }

    private fun updateAlternativeDate(isAlternative: Boolean) {
        binding.rdgAlternativeDate.check(if (isAlternative) R.id.rdYes else R.id.rdNo)
        binding.datesGroup.visibility = if (isAlternative) View.VISIBLE else View.GONE
    }

    private fun updatePlantingWindow(window: Int) {
        val planting = binding.cardPlanting
        planting.flexiblePlanting.isChecked = window > 0
        if (window > 0) {
            planting.rdgPlantingWindow.check(
                if (window == 1) R.id.rdPlantingOneMonth else R.id.rdPlantingTwoMonths
            )
        } else {
            planting.rdgPlantingWindow.clearCheck()
        }
    }

    private fun updateHarvestWindow(window: Int) {
        val harvest = binding.cardHarvest
        harvest.flexibleHarvest.isChecked = window > 0
        if (window > 0) {
            harvest.rdgHarvestWindow.check(
                if (window == 1) R.id.rdHarvestOneMonth else R.id.rdHarvestTwoMonths
            )
        } else {
            harvest.rdgHarvestWindow.clearCheck()
        }
    }

    private fun setupPlantingUI() = binding.cardPlanting.run {
        flexiblePlanting.setOnCheckedChangeListener { _, isChecked ->
            if (viewModel.alreadyPlanted.value == true) {
                showCustomWarningDialog(
                    getString(R.string.lbl_already_planted_title),
                    getString(R.string.lbl_already_planted_text)
                )
                flexiblePlanting.isChecked = false
                return@setOnCheckedChangeListener
            }
            rdgPlantingWindow.visibility = if (isChecked) View.VISIBLE else View.GONE
            if (!isChecked) viewModel.plantingWindow.value = 0
        }

        rdgPlantingWindow.setOnCheckedChangeListener { _, checkedId ->
            viewModel.plantingWindow.value = when (checkedId) {
                R.id.rdPlantingOneMonth -> 1
                R.id.rdPlantingTwoMonths -> 2
                else -> 0
            }
        }

        btnPickPlantingDate.setOnClickListener {
            showDatePicker(pickPlantingDate = true)
        }
    }

    private fun setupHarvestUI() = binding.cardHarvest.run {
        flexibleHarvest.setOnCheckedChangeListener { _, isChecked ->
            rdgHarvestWindow.visibility = if (isChecked) View.VISIBLE else View.GONE
            if (!isChecked) viewModel.harvestWindow.value = 0
        }

        rdgHarvestWindow.setOnCheckedChangeListener { _, checkedId ->
            viewModel.harvestWindow.value = when (checkedId) {
                R.id.rdHarvestOneMonth -> 1
                R.id.rdHarvestTwoMonths -> 2
                else -> 0
            }
        }

        btnPickHarvestDate.setOnClickListener {
            showDatePicker(pickHarvestDate = true)
        }
    }

    private fun setupAlternativeDateUI() {
        binding.rdgAlternativeDate.setOnCheckedChangeListener { _, checkedId ->
            viewModel.alternativeDate.value = checkedId == R.id.rdYes
        }
    }

    private fun setupButtons() = binding.twoButtons.run {
        btnFinish.setOnClickListener { validate(false) }
        btnCancel.setOnClickListener { closeActivity(false) }
    }

    override fun validate(backPressed: Boolean) {
        val plantingValid = !viewModel.plantingDate.value.isNullOrBlank()
        val harvestValid = !viewModel.harvestDate.value.isNullOrBlank()

        if (!plantingValid) {
            showCustomWarningDialog(
                getString(R.string.lbl_invalid_planting_date),
                getString(R.string.lbl_planting_date_prompt)
            )
            return
        }

        if (!harvestValid) {
            showCustomWarningDialog(
                getString(R.string.lbl_invalid_harvest_date),
                getString(R.string.lbl_harvest_date_prompt)
            )
            return
        }

        viewModel.saveSchedule(
            onSuccess = { closeActivity(backPressed) },
            onError = {
                Toast.makeText(this, it.message ?: "Error", Toast.LENGTH_SHORT).show()
                Sentry.captureException(it)
            }
        )
    }

    private fun showDatePicker(
        pickPlantingDate: Boolean = false,
        pickHarvestDate: Boolean = false
    ) {
        val picker = if (pickHarvestDate) {
            DateDialogPickerFragment.newInstanceForHarvest(viewModel.plantingDate.value ?: "")
        } else {
            DateDialogPickerFragment.newInstanceForPlanting()
        }

        picker.setOnDismissListener(object : IDatePickerDismissListener {
            override fun onDismiss(
                myCalendar: Calendar,
                selectedDate: String,
                plantingDateSelected: Boolean,
                harvestDateSelected: Boolean
            ) {
                myCalendar.timeInMillis
                if (pickPlantingDate) {
                    viewModel.updatePlantingDate(selectedDate)
                    binding.cardHarvest.lblSelectedHarvestDate.text = null
                } else if (pickHarvestDate) {
                    viewModel.updateHarvestDate(selectedDate)
                }
            }
        })

        picker.show(supportFragmentManager, DateDialogPickerFragment.TAG)
    }
}
